### ISSUE TITLE
`hacktoberfest` topic incompatible with manual PR merging

### ISSUE THREAD

**Issue by nightlark** [curl/curl#7865]
I suppose this is a bug with the `curl/curl` repo participating in hacktoberfest? Anyway...

### I did this

I'm not sure when the `hacktoberfest` topic was added to the repository, but when a recent PR got approved I received an email from Hacktoberfest saying a PR had been completed in the `curl/curl` repository. When the PR was closed by the manually merged commit method used here (i.e. PR shows with the red closed status icon rather than purple merged status icon), that overrides the approving review and the PR goes back to not counting towards `hacktoberfest`.

### I expected the following

I mentioned this to the folks running `hacktoberfest` as not following the valid PR logic they tell people, and the reasoning they gave for it working this way is that they think a PR with an approving review that has a closed status is more likely to mean that the project decided against making the change (and they don't have an easy way to track PRs that have been closed by a commit like this).

The suggested alternative by them for if the `curl` project wants to participate in `hacktoberfest` is labelling individual PRs with the `hacktoberfest-accepted` label, which in a test I did to confirm does appear to ignore the closed/merged status of PRs.

### curl/libcurl version

N/A

### operating system

N/A

**Comment by bagder** [comment]
So what do they think "closes #[num]" means when that [num] refers to a GitHub PR?

**Comment by bagder** [comment]
@mback2k [commented on twitter](https://twitter.com/mback2k/status/1449433645689802756) that "A PR also counts if it is first approved by a project member. So that would be the workaround here.". I sometimes skip the approve-step if I'm about to merge anyway, but I can certainly try to approve if I know it makes a difference to someone.

**Comment by bagder** [comment]
@nightlark can you link me the rules outlining these specifics?

**Comment by nightlark** [comment]
Theoretically https://hacktoberfest.digitalocean.com/resources/participation outlines the specifics:
```
A repository/project is considered to be participating in Hacktoberfest if the 'hacktoberfest' topic is present and is accepting public contributions via pull requests. An individual pull request can also be opted-in directly by adding the 'hacktoberfest-accepted' label.

A pull request is considered approved once it has an overall approving review from maintainers, or has been merged by maintainers, or has been given the 'hacktoberfest-accepted' label. A pull request with any label containing the word 'spam' or 'invalid' will be considered ineligible for Hacktoberfest.
```

However, on the HF Discord when I brought up the issue it was mentioned that there are nuances that aren't captured publicly: https://discord.com/channels/720788516837261322/735506302784110652/898301013243621426
```
Fwiw, there's a lot of nuance to what's considered valid that isn't captured there or publicly anywhere
But yes, a closed PR that only has an approval will not be considered valid
It was decided that it'd be more likely a PR had an approval and was then closed as not being accepted into the project, rather than having an approving review and then being closed because it was merged manually
```

**Comment by bagder** [comment]
The guidelines say:

> A pull request can be approved either by giving it an overall approving review from maintainers, or **by merging the pull request**, or adding the 'hacktoberfest-accepted' label 

Emphasis added by me. I think what is descriibed in this issue are added requirements not present on that page. We merge pull-requests that that. All.The.Time. And we've done it exactly like this for many years.

**Comment by MattIPv4** [comment]
👋 Just wanted to jump in here to say that whilst you are merging pull requests via Git, Hacktoberfest operates through GitHub and GitLab specifically, and so looks at what they report as being merged. When a PR on GitHub is merged manually, it is simply marked as closed on GitHub. The GitHub API tells us that it was closed, not merged. Unfortunately, as the API tells us that it was closed, that's how we treat it, we don't know it was manually merged.

(E.g. looking at https://github.com/curl/curl/pull/7862 via the API shows no indication it was merged, just that it was closed: https://api.github.com/repos/curl/curl/pulls/7862)

I understand that relying on the in-platform way of merging doesn't work for all projects, and therefore that part of Hacktoberfest's logic doesn't cater to all projects, and that's why the `hacktoberfest-accepted` label exists as an alternative way to mark a PR as valid for Hacktoberfest. Simply adding that label to any PR will qualify it for Hacktoberfest.

**Comment by bagder** [comment]
I think this is too intrusive having this "fun" label added on a per-PR basis. I rather not open the flood gates for that and allow other fun and games go down this route. If they cannot accept our merged PRs when clearly tagged in the git commit messages, I think we should officially remove the label and stay of out of this.

**Comment by bagder** [comment]
@nightlark if you have a PR we merged like this and you want to use this to apply for hacktoberfest rewards, please tell me which one it is and I'll label it like this for you as I don't want you as an individual contributor should suffer from this weirdo restriction they "punish" us with.

**Comment by MattIPv4** [comment]
I also hear your feedback on the site rules being too vague, around "by merging the pull request" and that approved PRs don't qualify if closed. Sorry that we didn't have more precise copy on there that reflected that we need the merges to be reported by the platforms Hacktoberfest operates on. I'll make sure we get the copy on the site updated to be more precise!

**Comment by MattIPv4** [comment]
Yes, that will be fine -- the `hacktoberfest-accepted` label will act to opt-in specific PRs, even if the project as a whole isn't opted in :)

**Comment by bagder** [comment]
Great, thanks! @nightlark let me know!

**Comment by bagder** [comment]
@MattIPv4 wrote:

> (E.g. looking at #7862 via the API shows no indication it was merged, just that it was closed: https://api.github.com/repos/curl/curl/pulls/7862)

That's almost like a GitHub bug. If you look at the page for https://github.com/curl/curl/pull/7862 it says `"bagder closed this in efffa66 [time] ago"` which means it was closed from a git commit message. That's at least something else than "just closed". I don't even see `efffa66` mention in that API response...

**Comment by nightlark** [comment]
@bagder no worries, I hadn't planned on PRs to curl being counted for HF, it was just the automated emails from HF that made me realize the repo had the `hacktoberfest` topic and that there was something strange going on with PRs closed by commit messages.

@MattIPv4 it would be really odd to use a keyword in a commit to close a PR if the commit wasn't making (or incorporating almost all) the same changes as the PR (equivalent to merging it in some form). This is a place where the most likely meaning of a closing keyword is different for a PR vs an Issue -- and a strong counter-signal to the `approved+closed=not merged` assumption.

If additional confirmation is needed for a check, the commit hash can be used to get the `author.login` and `author.id` fields (not committer) for a commit that should match those of the user who opened the PR: https://api.github.com/repos/curl/curl/commits/7759552b80c1b818454f8888b81eee23846294c9

**Comment by mback2k** [comment]
I could eventually manage to run a small but on the curl repo that automatically adds the label for PRs merged our way. What do you think?

**Comment by MattIPv4** [comment]
> should "spam" our project or PRs with labels for their benefit only

I don't really think this is for Hacktoberfest's benefit, beyond knowing that the PR was accepted. Projects opt into Hacktoberfest because they'd like folks from the community to participate in the project, and the label is then added to PRs by maintainers to ensure that those contributors are rewarded by Hacktoberfest. So really, it's to benefit the project and the contributors that are participating.

> And its not clear when we can remove the label(s) again.

Apologies if this wasn't apparent -- the final PRs can be submitted on Nov 1 12pm UTC (end of October in the last timezone), so those would mature 14 days later on Nov 15 12pm UTC, so any time after that it would be safe to remove the labels.

**Comment by mback2k** [comment]
> > And its not clear when we can remove the label(s) again.
> 
> Apologies if this wasn't apparent -- the final PRs can be submitted on Nov 1 12pm UTC (end of October in the last timezone), so those would mature 14 days later on Nov 15 12pm UTC, so any time after that it would be safe to remove the labels.

Okay, so I will also add a GitHub Actions workflow to remove all PR labels once Hacktoberfest is over, e.g. after November 15th.

@bagder I understand your concers, but do you think having these labels temporarily just during Hacktoberfest would really hurt that much? See for example https://github.com/home-assistant/core/pulls for a project that uses a lot of labels.

**Comment by bagder** [comment]
I think that many labels hurt, yes as it makes the important ones get lost in the forest. Does this single new label push us over the limit for "too many" ? Probably not.
